File: solutions/kotlin/difference-of-squares/1/src/main/kotlin/DifferenceOfSquares.kt
import java.util.stream.IntStream
import kotlin.math.pow

class Squares(val number: Int) {

    fun sumOfSquares() = sum { square(it) }

    fun squareOfSum() = square(sum { it })

    fun difference() = squareOfSum() - sumOfSquares()

    private fun sum(modifier: (Int) -> Int): Int {
        return IntStream.rangeClosed(1, number)
            .map(modifier)
            .sum()
    }

    private fun square(i: Int) = i.toFloat().pow(2).toInt()
}

File: solutions/go/difference-of-squares/1/difference_of_squares.go
package diffsquares

import "math"

// SquareOfSum calculates the square of the sum of the first given number
func SquareOfSum(n int) int {

	sum := 0
	for i := 1; i <= n; i++ {
		sum += i
	}

	return int(math.Pow(float64(sum), 2))
}

// SumOfSquares calculates the sum of the given first numbers squared
func SumOfSquares(n int) int {

	sum := 0
	for i := 1; i <= n; i++ {
		sum += int(math.Pow(float64(i), 2))
	}

	return sum
}

// Difference returns the difference between the square of sum
// and sum of squares calculated using the first given numbers
func Difference(i int) int {
	return SquareOfSum(i) - SumOfSquares(i)
}

